specs/phase1/beacon-chain.md

```diff
@@ -795,33 +805,45 @@ def process_attestation(state: BeaconState, attestation: Attestation) -> None:
     assert is_valid_indexed_attestation(state, get_indexed_attestation(state, attestation))
 
     # Process target, source, head, timeliness
-    flags_to_set = FLAG_SOURCE
+    flags_to_set = [FLAG_SOURCE]
     if data.target.root == get_block_root(state, data.target.epoch):
-        flags_to_set |= FLAG_TARGET
-    if data.beacon_block_root == get_block_root_at_slot(state, data.slot):
-        flags_to_set |= FLAG_HEAD
+        flags_to_set.append(FLAG_TARGET)
+        if data.beacon_block_root == get_block_root_at_slot(state, data.slot):
+            flags_to_set.append(FLAG_HEAD)
     if state.slot - data.slot == 1:
-        flags_to_set |= FLAG_VERY_TIMELY
+        flags_to_set.append(FLAG_VERY_TIMELY)
     if state.slot - data.slot <= integer_squareroot(SLOTS_PER_EPOCH):
-        flags_to_set |= FLAG_TIMELY
+        flags_to_set.append(FLAG_TIMELY)
 
     flags = (state.current_epoch_reward_flags if is_current_epoch_attestation else state.previous_epoch_reward_flags)
 
+    # Reward proposer for new attestations
     for participant in get_attesting_indices(state, data, attestation.aggregation_bits):
         active_position = get_active_validator_indices(state, data.target.epoch).index(participant)
-        flags[active_position] |= flags_to_set
+        if not flags[active_position][FLAG_SOURCE]:
+            increase_balance(state, get_beacon_proposer_index(state), get_proposer_reward(state, participant))
+
+    # Update participation flags
+    for participant in get_attesting_indices(state, data, attestation.aggregation_bits):
+        active_position = get_active_validator_indices(state, data.target.epoch).index(participant)
+        for flag in flags_to_set:
+            flags[active_position][flag] = True
```

Comment:
Why not merge the above two loops? Right before the line `flags[active_position][flag] = True`, add the `if not flags[active_position][FLAG_SOURCE]:` clause from above.